a08.evaluator/Evaluator.cs

```
@@ -38,31 +45,30 @@ class Evaluator {
 
    void Process (Token token) {
       switch (token) {
-         case TNumber num: 
-            mOperands.Push (num.Value); 
+         case TNumber num:
+            mOperands.Push (num.Value);
             break;
          case TOperator op:
-            while (mOperators.Count > 0 && mOperators.Peek ().Priority > op.Priority)
-               ApplyOperator ();
+            while (mOperators.Count != 0 && mOperands.Count != 0 && mOperators.Peek ().Priority >= op.Priority) ApplyOperator ();
             mOperators.Push (op);
             break;
-         case TPunctuation p:
-            BasePriority += p.Punct == '(' ? 10 : -10;
+         case TPunctuation p when p.Punct is '(':
             break;
-         default:
-            throw new EvalException ($"Unknown token: {token}");
+         case TPunctuation p when p.Punct is ')':
+            ApplyOperator (); break;
       }
    }
    readonly Stack<double> mOperands = new ();
    readonly Stack<TOperator> mOperators = new ();
 
    void ApplyOperator () {
       var op = mOperators.Pop ();
-      var f1 = mOperands.Pop ();
-      if (op is TOpFunction func) mOperands.Push (func.Evaluate (f1));
-      else if (op is TOpArithmetic arith) {
-         var f2 = mOperands.Pop ();
+      double f1 = mOperands.Pop (), f2;
+      if (op is TOpArithmetic arith) {
+         try {
+            f2 = mOperands.Pop ();
+         } catch (Exception) { return; }

```

If you want to use try catch, you should not just return. It is better for application to crash instead of silently ignoring possible unexpected behavior. Also, you should log the exception. In our case, don't use try catch. Catch the exception at the main method.

---

In this case it was proving useful for when there are too many operators and it throws a stack empty exception, I wanted it to simply return to the previous method and throw the relevant exception if its not handled. Should I throw it here itself?

---

Then catch that specific exception and throw your own exception or other method to handle.

---

Can I log it and return? or should I log it and rethrow?

---

Also, should I just log just the system exceptions or should I log every exception that I'm throwing too?